src/Select.purs

```diff
@@ -22,124 +25,94 @@ import Effect.Ref (Ref)
+  = GetVisibility (Visibility -> a)
   | ReplaceItems (Array item) a
-  | Raise (o Unit) a
-  | Initialize a
-  | Receive (Input o item) a
-
-type Query o item = Free (QueryF o item)
-
--- | Trigger the relevant action with the event each time it occurs
-always :: ∀ a b. a -> b -> Maybe a
-always = const <<< Just
-
--- | Perform a new search with the included string.
-search :: ∀ o item. String -> Query o item Unit
-search s = liftF (Search s unit)
-
--- | Change the highlighted index to the next item, previous item, or a
--- | specific index.
-highlight :: ∀ o item. Target -> Query o item Unit
-highlight t = liftF (Highlight t unit)
-
--- | Triggers the "Selected" message for the item at the specified index.
-select :: ∀ o item. Int -> Query o item Unit
-select i = liftF (Select i unit)
-
--- | Trigger the DOM focus event for the element we have a reference to.
-triggerFocus :: ∀ o item . Query o item Unit
-triggerFocus = liftF (Focus true unit)
-
--- | Trigger the DOM blur event for the element we have a reference to
-triggerBlur :: ∀ o item . Query o item Unit
-triggerBlur = liftF (Focus false unit)
-
--- | Register a key event. `TextInput`-driven components use these only for
--- | navigation, whereas `Toggle`-driven components also use the key stream for
--- | highlighting.
-key :: ∀ o item . KE.KeyboardEvent -> Query o item Unit
-key e = liftF (Key e unit)
-
--- | A helper query to prevent click events from bubbling up.
-preventClick :: ∀ o item . ME.MouseEvent -> Query o item Unit
-preventClick i = liftF (PreventClick i unit)
-
--- | Set the container visibility (`On` or `Off`)
-setVisibility :: ∀ o item . Visibility -> Query o item Unit
-setVisibility v = liftF (SetVisibility v unit)
-
--- | Get the container visibility (`On` or `Off`). Most useful when sequenced
--- | with other actions.
-getVisibility :: ∀ o item . Query o item Visibility
-getVisibility = liftF (GetVisibility identity)
-
--- | Toggles the container visibility.
-toggleVisibility :: ∀ o item . Query o item Unit
-toggleVisibility = getVisibility >>= not >>> setVisibility
-
--- | Replaces all items in state with the new array of items.
-replaceItems :: ∀ o item . Array item -> Query o item Unit
-replaceItems items = liftF (ReplaceItems items unit)
-
--- | A helper query that the component that mounts `Select` can use to embed its
--- | own queries. Triggers an `Emit` message containing the query when triggered.
--- | This can be used to easily extend `Select` with more behaviors.
-raise :: ∀ o item . o Unit -> Query o item Unit
-raise o = liftF (Raise o unit)
-
--- | Sets the component with new input.
-receive :: ∀ o item . Input o item -> Query o item Unit
-receive i = liftF (Receive i unit)
-
--- | Initializes the component on mount.
-initialize :: ∀ o item. Query o item Unit
-initialize = liftF (Initialize unit)
+  -- send a query through `Select` to a child component
+  | SendQuery (ChildQueryBox ps (Maybe a))
+  | PerformAction (Action item v ps m) a
```

Comment:
`PerformAction` makes all actions accessible from the parent via `H.query`. Not sure if this is desirable; perhaps some subset of actions could be split out and made accessible while others are private (receive, initialize, preventClick, key, andThen come to mind as ones that don't have much use being called from a parent).